### PR TITLE
Annotate K8s Pods with flow name, step, run_id for ddog integration

### PR DESCRIPTION
**Pull request by talebzeghmi**
- prefix labels with `metaflow.org/`

example pod annotation with run:
```bash
python resources_flow.py kfp run --experiment mf_test --tag t1
```
```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    metaflow.org/experiment: mf_test
    metaflow.org/flow_name: ResourcesFlow
    metaflow.org/run_id: kfp-1047f250-6d8d-441b-b81e-285abe19620e
    metaflow.org/step: start
    metaflow.org/tag_aip-metaflow-sandbox: "true"
    metaflow.org/tag_mf_test: "true"
    metaflow.org/tag_t1: "true"
```

### CODE REVIEW

**Comment by talebzeghmi** [comment]
I've switch from annotations to labels which is more correct:
> Labels can be used to organize and to select subsets of objects
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/

File: metaflow/plugins/kfp/tests/.gitlab-ci.yml [hunk 1]
```diff
@@ -163,11 +162,10 @@ test:nonprod:
...
         ${BUILT_IMAGE_FULL_PATH}
         bash -c "
-          export KFP_RUN_URL_PREFIX=https://kubeflow.corp.stage-k8s.zg-aip.net/ && 
-          export KFP_SDK_NAMESPACE=metaflow-integration-testing-dev && 
-          export METAFLOW_DATASTORE_SYSROOT_S3=s3://aip-example-stage/metaflow/ && 
-          export METAFLOW_USER=${GITLAB_USER_EMAIL} &&
-          cd /home/zservice/metaflow/metaflow/plugins/kfp/tests && 
+          export KFP_RUN_URL_PREFIX=https://kubeflow.corp.stage-k8s.zg-aip.net/ &&
+          export KFP_SDK_NAMESPACE=metaflow-integration-testing-dev &&
+          export USER=${GITLAB_USER_EMAIL} &&
```
Comment (cloudw) [inline metaflow/plugins/kfp/tests/.gitlab-ci.yml:167]: I recognize the where the datastore change come from, but I am not quite sure what caused changes from `METAFLOW_USER` to `USER`. Are they both from your changes in kubeflow namespace (profile)?

**Review by cloudw — COMMENTED** [review]
Look good to me. Just waiting on resolutions to @alexlatchford 's comments before approving.

File: metaflow/plugins/kfp/tests/.gitlab-ci.yml (hunk 1 above)
Comment (talebzeghmi) [reply metaflow/plugins/kfp/tests/.gitlab-ci.yml:167]: Thanks for asking.  If it is METAFLOW_USER then the run gets tagged as that user, in this case the GITLAB_USER_EMAIL.  see [code](https://github.com/zillow/metaflow/blob/88a639649d818c194d26824e706641cde63bf7f6/metaflow/plugins/kfp/kfp.py#L716-L720).

It's cleaner if the run is a `kfp-user` (an automated run 

USER doesn't persist and would then only be used for creating the pipeline see [code](https://github.com/zillow/metaflow/blob/88a639649d818c194d26824e706641cde63bf7f6/metaflow/plugins/kfp/kfp.py#L157-L165)

File: metaflow/plugins/kfp/kfp.py [hunk 2]
```diff
@@ -532,6 +536,21 @@ def _set_container_settings(container_op: ContainerOp, kfp_component: KfpCompone
...
+        container_op.add_pod_label(f"{prefix}/flow_name", self.name)
+        container_op.add_pod_label(f"{prefix}/step", node.name)
+        container_op.add_pod_label(f"{prefix}/run_id", metaflow_run_id)
+
+        if self.experiment:
+            container_op.add_pod_label(f"{prefix}/experiment", self.experiment)
+        if self.tags and len(self.tags) > 0:
+            for tag in self.tags:
+                annotation_name = f"{prefix}/tag_{tag}"
+                container_op.add_pod_label(annotation_name, "true")
```
[the remark this answers is not in the document]
Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:552]: Thanks! I shared it with him before coding.

**Comment by talebzeghmi** [comment]
>Looks good to me but I have a higher level question. What are your expectations for what customers can do with the workflows with these tags?

>I'm guessing they can use the Metaflow CLI tools (and potentially the newer foundry CLI/SDK) to retrieve their runs, is there any expectations these tags would allow filtering from Splunk or Datadog (or elsewhere)?

These tags would be used to organize flows and `--experiment` (kfp plugin feature only) as outlined in MF [docs](https://docs.metaflow.org/metaflow/tagging).

I'm adding the tags as labels to K8s in the hope that we could use those for cost attribution.  I don't see how they would be used by datadog yet, and not the highest priority for datadog.  The main scenario for ddog is to view the CPU/mem/etc resources on a Flow or Flow Step.  @alexlatchford

**Comment by alexlatchford** [comment]
> I'm adding the tags as labels to K8s in the hope that we could use those for cost attribution. I don't see how they would be used by datadog yet, and not the highest priority for datadog. The main scenario for ddog is to view the CPU/mem/etc resources on a Flow or Flow Step. @alexlatchford

Thanks for confirming I think you'd need to conform to [this spec](https://docs.google.com/document/d/1nAYxebAhsK_X8fswds9DPV0xbcPzHJ2Wb9nce3oAsTc/edit#heading=h.6a09xyea7k34) to get the labels working for cost attribution (at least the ZGCP solution Ledger) so I think you'd also need something like `tags.ledger.zgtools.net/experiment-name` as they have a specific format we need to conform to for that.

Yeah I tend to agree with lower priority for the Datadog monitoring but even to get the tags for `Flow` or `FlowStep` to flow into Datadog you'd need to either use their Tag Auto-Discovery feature I think which would be needing to fill in a `ad.datadoghq.com/tags` annotation with a value like `'{"flow_name": "...", "flow_step_name": "..."}'` as they cannot pull in just any labels or annotations unfortunately 😒

**Comment by talebzeghmi** [comment]
> tags.ledger.zgtools.net/experiment-name

@alexlatchford these are zillow specific and could be handled by the k8s `PodDefault`.

>ad.datadoghq.com/tags annotation with a value like '{"flow_name": "...", "flow_step_name": "..."}'

I suggest as George has to update the Datadog agent daemonset configuration
`DD_KUBERNETES_POD_LABELS_AS_TAGS='{"metaflow.org/*":"%%label%%"}'`

If these options will not work, then we can find another solution, hopefully a Zillow specific plugin independent of this code base.

**Pull request merged**